docs/architecture/adr-038-state-listening.md

```diff
@@ -0,0 +1,584 @@
+	...
+
+	// ListeningEnabled returns if listening is enabled for the KVStore belonging the provided StoreKey
+	ListeningEnabled(key StoreKey) bool
+
+	// SetListeners sets the WriteListeners for the KVStore belonging to the provided StoreKey
+	// It appends the listeners to a current set, if one already exists
+	SetListeners(key StoreKey, listeners []WriteListener)
+}
+```
+
+```go
+type CacheWrap interface {
+	...
+
+	// CacheWrapWithListeners recursively wraps again with listening enabled
+	CacheWrapWithListeners(listeners []WriteListener) CacheWrap
+}
+
+type CacheWrapper interface {
+	...
+
+	// CacheWrapWithListeners recursively wraps again with listening enabled
+	CacheWrapWithListeners(listeners []WriteListener) CacheWrap
+}
+```
+
+### MultiStore implementation updates
+
+We will modify all of the `Store` and `MultiStore` implementations to satisfy these new interfaces, and adjust the `rootmulti` `GetKVStore` method
+to wrap the returned `KVStore` with a `listenkv.Store` if listening is turned on for that `Store`.
+
+```go
+func (rs *Store) GetKVStore(key types.StoreKey) types.KVStore {
+	store := rs.stores[key].(types.KVStore)
+
+	if rs.TracingEnabled() {
+		store = tracekv.NewStore(store, rs.traceWriter, rs.traceContext)
+	}
+	if rs.ListeningEnabled(key) {
+		store = listenkv.NewStore(store, rs.listeners[key])
+	}
+
+	return store
+}
+```
+
+We will also adjust the `cachemulti` constructor methods and the `rootmulti` `CacheMultiStore` method to forward the listeners
+to and enable listening in the cache layer.
+
+```go
+func (rs *Store) CacheMultiStore() types.CacheMultiStore {
+	stores := make(map[types.StoreKey]types.CacheWrapper)
+	for k, v := range rs.stores {
+		stores[k] = v
+	}
+	return cachemulti.NewStore(rs.db, stores, rs.keysByName, rs.traceWriter, rs.traceContext, rs.listeners)
+}
+```
+
+### Exposing the data
+
+We will introduce a new `StreamingService` interface for exposing `WriteListener` data streams to external consumers.
+
+```go
+// StreamingService interface for registering WriteListeners with the BaseApp and updating the service with the ABCI context
+type StreamingService interface {
+	Listeners() map[sdk.StoreKey][]storeTypes.WriteListener // returns the streaming service's listeners for the BaseApp to register
+	BeginBlockReq(req abci.RequestBeginBlock) // update the streaming service with the latest RequestBeginBlock message
+	BeginBlockResres abci.ResponseBeginBlock) // update the steaming service with the latest ResponseBeginBlock message
+	EndBlockReq(req abci.RequestEndBlock) // update the steaming service with the latest RequestEndBlock message
+	EndBlockRes(res abci.ResponseEndBlock) // update the steaming service with the latest ResponseEndBlock message
+	DeliverTxReq(req abci.RequestDeliverTx) // update the steaming service with the latest RequestDeliverTx message
+    DeliverTxRes(res abci.ResponseDeliverTx) // update the steaming service with the latest ResponseDeliverTx message
```

Comment:
We're adding more computation to be run on each ABCI call. It would be good to have an idea of the performance overhead.

Reply:
Completely agree and I admittedly do not have a great feel for this yet, I could do some preliminary benchmarking. These methods would be updating the `StreamingService` with the latest ABCI message, so that it can e.g. generate a new file which retains that context and which we write out the state changes that are caused by that particular message.